pennylane/wires.py

```diff
@@ -0,0 +1,399 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
+"""
+from collections import Sequence, Iterable
+from numbers import Number
+import numpy as np  # for random functions
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+class Wires(Sequence):
+    """
+    A bookkeeping class for wires, which are ordered collections of unique objects. The i'th object represents
+    the i'th quantum subsystem.
+
+    There is no conceptual difference between registers of multiple wires and single wires,
+    which are just wire registers of length one.
+
+    Indexing and slicing this sequence will return another ``Wires`` object.
+
+    Args:
+         wires (Any): If iterable, interpreted as an ordered collection of unique objects representing wires.
+            Any other input type is converted into an iterable of a single entry, and hence interpreted as a
+            single wire.
+    """
+
+    def __init__(self, wires):
+
+        if isinstance(wires, Wires):
+            # If input is already a Wires object, just copy its wire list
+            self.wire_list = wires.wire_list
+
+        elif isinstance(wires, Iterable):
+            # If input is an iterable, check that entries are unique and convert to a list
+            try:
+                if len(set(wires)) != len(wires):
+                    raise WireError(
+                        "Wires must be unique; got {}.".format(wires)
+                    )
+            except TypeError:
+                raise WireError("Cannot create wires from elements of type {}.".format(type(wires[0]))) # TODO: edge case that iterable contains different types
+
+            self.wire_list = list(wires)
+
+        else:
+            # If input is not an iterable, interpret it as an object representing a single wire
+            self.wire_list = [wires]
+
+    def __getitem__(self, idx):
+        """Method to support indexing. Returns a Wires object representing a register with a single wire."""
+        return Wires(self.wire_list[idx])
+
+    def __len__(self):
+        """Method to support ``len()``."""
+        return len(self.wire_list)
+
+    def __contains__(self, item):
+        """Method checking if Wires object contains an object."""
```

Comment:
> ```pycon
> >>> w = Wires([0,1,2])
> >>> assert 0 in w
> ```

This would be super useful! Out of curiosity, is this already supported by default due to inheriting from `Sequence`? Or do we need to define it ourselves?

Reply:
Good point, I will comment in the new PR.